### ISSUE TITLE
[Snapp] Optimistic: Rollback account state on successful dispute

### ISSUE THREAD

**Issue by bh2smith** [gnosis/dex-contracts#89]
# Rollback on Successful Dispute

Because this trading engine offers decentralized order matching, dispute-resolution mechanism must be implemented.
The *anchor contract* is incapable of determining valid from invalid solution proposals and assumes validity by default.
When a challenge is successful, there must be possibility to rollback or, at least, ignore faulty state transitions.

The smart contract only knows a compressed representation of the account balances on the exchange at any given time, and thus, is incapable of determining when a balance update (i.e. state transition is valid).

## Setting the stage

In the current (MVP) implementation of the `SnappBase` contract, account states are stored as an array of hashes 

```
bytes32 [] stateRoots;
```

There are three types of state transition (Deposits, Withdraws and Auction Settlement): All different transition types are similar in the sense that they are the result of processing a batch collected requests. However, no two of them are treated identically must be considered individually for implementation dispute resolution. Below is a brief summary of each transition type with their subtle differences highlighted**!**

1. **Deposit**
	 - Deposit requests are collected in a batch via `requestDeposit`. 
	 - **! Funds are immediately transferred to the contract upon request. !** 
	 - Sender's new account balance is only represented `stateRoot` when batch of requests has been processed (via `applyDeposits`). 
	This process suggests that deposit request batches can not be forgotten in the event of a rollback
2. **Withdraw**
	 - Withdraw requests are collected in a batch via `requestDeposit`. 
	 - Funds are **NOT** taken out of the account
	 - Sender's new account balance is only represented `stateRoot` when batch of requests has been processed (via `applyWithdrawals`).
	 - **! Funds are only transferred out of the contract after the Withdrawals have been applied ( via `claimWithdraw`) !**. [This will imply that withdraw transitions must be finalized before funds can be claimed].

3. **Auction Settlement**
	- limit orders are collected in a batch via `placeOrder`
	- orders are put in to an optimization problem and solutions are proposed by many solvers, collected by contract and organized in the form of a max heap.
	- after solution proposal period has ended, contract selects the "best" by popping the proposal heap. 
		* Note: if there are no solutions, the `NullSolution` is applied.
	- proposer to provided solution information and new state hash (via `applyAuction`)

## Finalization

The simplest imaginable form of state finalization is to impose a fixed time-delay `T` (aka challenge period) for each state update. Thus, disabling any additional updates until the challenge period for the current state has passed. Such a finalization is agnostic to the transition types and allows for one to carefully consider the implementation details of rollback individually without any additional complexity.

An alternative, slightly more complex, might be to consider the definition of finalization differently for each individual transition type. This idea will be investigated after giving implementation proposals for the simple case. 

## Challenge & Resolution

One key factor in dispute resolution is the number of transactions required for the contract to verify if rollback is required. This essentially breaks down into single vs. multiple transaction validation. For example, based on the discussions in #87, we see that snark-resolution could be validated in a "single" transaction (if the deployment of proof contract is considered separately), whereas on-chain validation often requires multiple transactions (due to payload constraints and overall block-gas limit). When a fraud proof can be validated in a single transaction, it suffices to implement the challenge, validation and rollback all within the single function. However, for multi-transaction fraud-proof validation, roll-back will requires a few steps such as

- challenge state by index and raise flag.
- provide pieces of proof in steps.
- if state turns out invalid (after several steps), rollback.

The multi-transaction case inevitably implies that the time-delay must be reset (or extended in some way) on each successful step in the fraud-proof (giving the prover sufficient time to continue). In this case one may want to consider the possibility of allowing someone to provide a legitimacy proof to terminate the dispute process. Such time extension could also result in a DOS attack if not implemented properly.

It will likely turn out, that a hybrid solution for rolling back both single and multi transaction validation. This means consideration in both scenarios. Notice also, that any single transaction validation may not even require the challenger to bond their proof.

**Comment by bh2smith** [comment]
# Single Transaction Validation with Fixed Time Delay

For now, we explore the simple implementation of rolling back a state transition in each of the three separate cases.

In all cases, the current state root is deleted, thus leaving the previous as the current. That is, 

```js
delete stateRoots[stateRoots.length() - 1];
```
Depending on implementation, this could leave the previous state open to challenge once again (if time-delay is reset on rollback). **It will become evident in the discussion of withdraw rollback, that the exposed previous state should no longer be disputable after rollback.**

Furthermore, at the very least, there would be a generic event of the form 

```
rollbackState(transitionType, batchIndex)
```

It will become more clear whether special events are necessary

## Rollback Deposit

The deposit requestor has already transferred the funds into the contract meaning the `PendingFlux` (batch information) **can not** be forgotten. This means, the rollback can't purge the deposit batch, but must nullify the `appliedAccountStateIndex` so to revert the batch into a state of *unapplied*.

```js
fn rollbackDeposit(
    uint batchIndex, 
    bytes fraudProof,
    uint bond  // may not be necessary in the single transaction case.
) public {
    // Check that current state index is a result of the challenged deposit.
    require(deposits[index].appliedAccountStateIndex == stateIndex);

    if validateDepositProof(PendingFlux deposits[index], fraudProof) {
        delete stateRoots[stateRoots.length() - 1];    // roll-back state
        deposits[index].appliedAccountStateIndex = 0;  // revert batch to unprocessed.
        emit DepositRollBack(batchIndex);
    }
}
```

## Rollback Withdraw

Unlike deposits, since no funds are actually transferred on in or out of the contract on withdraw request, it doesn't actually matter if the processed batch information is reverted into a state of unapplied or forgotten. However, from a user perspective and at no cost, this can be implemented without such an inconvenient loss. This suggests `rollbackWithdraw` can be implemented analogous to deposits.

**Note that,** processed withdraws cannot be claimed until the corresponding withdraw transition has been finalized. This follows because withdraw claims transfer value out of the contract which can not be reverted. 

This suggests that successful rollback of state should not leave the previous state exposed to dispute. That is, finalization is defined purely by the time delay, and not by the fact that it is the "current state"


## Rollback Auction

In the (decentralized version of our) current implementation, `applyAuction` is called by the chosen solution proposer. It contains all information relevant to the trade execution included in the transaction's payload) and updates the `stateRoot` according to the sender's proposed new value.

In this case, rolling back amounts to reverting state (as in both of the above cases), revert the processed order batch by nullifying `appliedAccountStateIndex` while alternatively choosing the next best solution in the heap. This should then announce or update the value of "chosen proposer" and reset the submission timer.

**Comment by fleupold** [comment]
> This process suggests that deposit request batches can not be forgotten in the event of a rollback

Technically, people could locally verify if a state is valid before it is even challenged on-chain. Therefore, we could argue that depositing into an invalid state is an invalid action and may lead to loss of funds. Thus it's not strictly required to re-apply deposits after rollback.

> The multi-transaction case inevitably implies that the time-delay must be reset 

Could you elaborate why this is inevitable? If the time-delay was something like 2 hours, couldn't we expect event the multi-transaction challenge to be completed by then?

> In all cases, the current state root is deleted, thus leaving the previous as the current.

If we consider a batch time of 3 minutes but a finalization time of 2 hours, it's possible that new states have been added to the list since the invalid state transition was applied. These new transitions might in themselves be valid and thus not challengeable. I believe we therefore need a way to challenge an arbitrary index and not just the most recent one.

What would be the implications of this to your solutions?

> This should then announce or update the value of "chosen proposer" and reset the submission timer.

I'm a bit afraid that an attacker could submit the `n` highest bids and then halt the exchange for `n * submission period` minutes. Alternatively, we could at this point allow anyone (e.g. in the top 10) to submit a solution. But this question is orthogonal to the rollback mechanism since it already applies to decentralized solution collection without rollbacks.

**Comment by bh2smith** [comment]
> Technically, people could locally verify if a state is valid before it is even challenged on-chain. Therefore, we could argue that depositing into an invalid state is an invalid action and may lead to loss of funds. Thus it's not strictly required to re-apply deposits after rollback.

Consider the following scenario; One could deposit at a time where the state is perfectly valid (transferring their funds to the contract), that batch is processed incorrectly and this state transition is rolled back. If that batch of deposits is forgotten then the contract has essentially stolen the funds.

This implies deposit batches can not be forgotten on rollback

**Comment by bh2smith** [comment]
> Could you elaborate why this is inevitable? If the time-delay was something like 2 hours, couldn't we expect event the multi-transaction challenge to be completed by then?

Time delay was never expected to be on the magnitude of two hours. I was thinking something in the range of minutes. Perhaps the same as the batch collection times. Lets go with 3 minutes (this is why it would need to be reset).

> If we consider a batch time of 3 minutes but a finalization time of 2 hours, it's possible that new states have been added to the list since the invalid state transition was applied. These new transitions might in themselves be valid and thus not challengeable. I believe we therefore need a way to challenge an arbitrary index and not just the most recent one.

Similarly, this time delay is expected to be something reasonable. This investigation does not yet touch on the point of rolling back a specified index. In that is described with "time-delay" all states which are not current are considered finalized.

An alternate description where this time delay is specific to the transition-type (say 1, 2 and 3 minute delays for finalization of deposits withdrawals and auctions) could also be manageable with minimal complexity increase.

I would think that we need to make very careful considerations to be allowed to rollback at a specific index. Meanwhile some appropriate definition of finalization must be involved.

**Comment by fleupold** [comment]
I think @josojo was thinking about a time-delay of up to one week (which would allow people to come online at least once within the time and verify that their funds are still secure). I believe this is too long as we cannot reclaim funds until then.

However, 3 minutes might be too short to detect a wrong transition, create the proof (in case of snarks this could take up to 30 minutes) and find a block to include this large transaction in. Even without snarks, I'd like to see this period to be more like 30 minutes (with snarks more like 2 hours).

**Comment by bh2smith** [comment]
In the event that we need time-delays on the order of 30 minutes or more, I would completely agree that we can not hold back state transitions by the time-delay and would have to allow transition on non-finalized states. This is a good argument leading into the need for **forkable states**.

**Comment by fleupold** [comment]
## Insight on Forking vs Rollback

I was going to write down the alternative suggestion of state dispute, which we referred to a forking. With each challenge, the challenger would basically provide an alternative state hash which in their mind represents the correct state transition. This would create a fork in the chain of states:

```
           |--> state*_i+1 (submitted by challenger)
state_i  --|
           |--> state_i+1 (submitted by original author)
```

People in the network could then locally verify which of the two is correct and keep submitting future interactions (orders, deposits, withdrawals) on the subchain that they think is correct.

This approach seemed fruitful in a challenge/response scenario, where the original solution submitter was given a response time (e.g. another 2 hours) to generate and provide the validity proof. Without a forking system, the exchange would have to be halted during that time.

I believe however, that in the scenario where the challenger can provide the fault proof themselves, forking and rollback become the same thing (except that in forking the challenger proposes a new state as well).

This is because there will never actually be more than one branch. If the fraud proof is valid, the original branch will be deleted immediately. Otherwise the suggested alternative branch is deleted immediately.
It would still make some sense in the case of a multi-transaction challenge, however we can assume that the number of transaction would still be small (<10) and thus the time from start to finish would only be minutes rather than hours.

Forking could also be further explored in a scenario where we don't want to rely on dispute resolution with immediate finality (using snarks or solidity) but instead rely on a fork-choice rule (e.g. longest subchain)

**Comment by bh2smith** [comment]
> I'm a bit afraid that an attacker could submit the n highest bids and then halt the exchange for n * submission period minutes. Alternatively, we could at this point allow anyone (e.g. in the top 10) to submit a solution. But this question is orthogonal to the rollback mechanism since it already applies to decentralized solution collection without rollbacks.

Yet another very good point and a good indicator that forkable states could come in very handy. However suppose the attack actually had a feasible solution and they submit `n` proposals with each of the first `n-1` being garbage and the last being legit. They would still hold up the exchange by that same time delay... since their last solution would be "legit".

I wonder if it would make sense to allow someone (i.e. anyone) to submit feasible solutions that are greater than the proposed surplus in order to avoid such a hold up. This could be very interesting to look into.

**Comment by bh2smith** [comment]
With regards to 

> Insight on Forking vs Rollback

I can imagine a situation where providing an alternative state proposal upon challenge can be implemented for deposits and withdrawals (since the information is available and the transition is uniquely determined by the request data). Unfortunately, however, I do not believe that this can work for auction settlement since there is an account which is "next in line" to be the expected submitter. 

Although, this brings to mind the previous comment mentioned briefly that

> I wonder if it would make sense to allow someone (i.e. anyone) to submit feasible solutions that are greater than the proposed surplus in order to avoid such a hold up. This could be very interesting to look into.

Only question remains is "who would pick up the reward in such a scenario?"

**Comment by fleupold** [comment]
That's an interesting thought. I believe we should discuss this in a separate issue as part of how we implement decentralised solution submission (ie. competing for solution), since the problem roots there.
Even without any rollbacks, how do we make sure somebody claiming to have the top n solutions doesn't spend ages submitting them.

Back to the questions of this task. How would your proposed model change assuming the `finalization period >> batch time`?

**Comment by josojo** [comment]
>  I believe we should discuss this in a separate issue as part of how we implement decentralized solution submission (ie. competing for solution), since the problem roots there.

Yeah, this question is fundamental and likely first needs to be solved, before fixxing rollbacks.
I would argue that we should not protect our exchange from downtime of n invalid solutions, beyond a big bond. The reason is that it is likely anyways cheaper to DDOS our exchange by submitting spam orders and buy up the rent accounts.

**Comment by bh2smith** [comment]
When finalization time is significantly longer than batch time (i.e. say even 10-fold) then the contract will have to allow state transition on unfinalized state roots and thus roll back will have to be capable of handling revert over several transitions. THe below image describes a rollback to a state `s_2` after a bad auction `A_1` transition occurs even after several other transitions had already been processed. 
```
     D_1        A_1         W_1        A_2        D_2        W_2
s_1 -----> s_2 -----> s_3* -----> s_4 -----> s_5 -----> s_6 -----> s_7 
```
Suppose that s_3 was a successfully challenged (only after several other transations had occured)

Note that A, D, W represent state transitions of type Auction, Deposit and Withdraw respoectively.

In this case, the contract would roll back to state `s_2` (i.e. delete the states `s_i` for `i > 3` from the `stateRoots` array `s_2` the new "current state") while setting the `PendingFlux.appliedStateIndex = 0` for each of `W_1`, `W_2` and `D_2`. Furthermore, the orders collected in any auctions beyond `A_1` would simply be forgotten (i.e. resetting both the order's `appliedStateIndex` and `shaHash` to 0). 

**Note that**
-  Proposed deletion of order batches may require special consideration for Rollover orders.
- We must analyze and evaluate the worst-case gas costs for looping back over old state transitions while resetting to older states (could cause and out-of-gas type attack rollback)
- This proposal does not involve any type of forking (introducing such complexity should not be necessary), as challenges should not take a long time. At this moment we assume that even a challenge which can take up to 10 transactions wouldn't necessarily take a long time (even if on chain).

- To summarize, current idea is to reset deposit and withdraw batches to "unapplied", and forget orders (in the event that several state transitions are rolled back).
